### PR TITLE
Add support for dynamic loading.

### LINKED ISSUE #3: Support dynamic loading
https://github.com/opentracing/opentracing-cpp/issues/28

### PR DESCRIPTION
**Pull request by rnburn**
Implements the dynamic loading support available in opentracing-cpp-1.3.

Fixes #3

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -283,8 +287,14 @@ if(BUILD_TESTING)
...
   target_link_libraries(
-      UnitTest testutils GTest::main jaegertracing-static ${LIBS})
+      UnitTest testutils GTest::main jaegertracing ${LIBS})
   add_test(NAME UnitTest COMMAND UnitTest)
+
+  add_executable(DynamicallyLoadTracerTest src/jaegertracing/DynamicallyLoadTracerTest.cpp)
+  target_link_libraries(
+    DynamicallyLoadTracerTest OpenTracing::opentracing-static)
+  add_test(NAME DynamicallyLoadTracerTest COMMAND DynamicallyLoadTracerTest 
+           ${CMAKE_BINARY_DIR}/${CMAKE_SHARED_LIBRARY_PREFIX}jaegertracing${CMAKE_SHARED_LIBRARY_SUFFIX})
```
Comment (isaachier) [inline CMakeLists.txt]: Can you change `${CMAKE_BINARY_DIR}` with `${CMAKE_CURRENT_BINARY_DIR}`. I think it is better for recursive builds.

File: src/jaegertracing/DynamicLoad.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+#include <system_error>
+
+#include <opentracing/dynamic_load.h>
+
+#include "TracerFactory.h"
+#include "jaegertracing/Tracer.h"
+
+int OpenTracingMakeTracerFactory(const char* opentracingVersion,
+                                 const void** errorCategory,
+                                 void** tracerFactory)
```
Comment (isaachier) [inline src/jaegertracing/DynamicLoad.cpp]: Can you add `assert(tracerFactory != nullptr)`?

File: src/jaegertracing/DynamicLoad.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+{
+    if (std::strcmp(opentracingVersion, OPENTRACING_VERSION) != 0) {
```
Comment (isaachier) [inline src/jaegertracing/DynamicLoad.cpp]: I'm going to assume this means the OpenTracing C++ API version. I see you use the version string to generate it. Are you sure we want to use minor and patch versions, or just major version?

File: src/jaegertracing/DynamicallyLoadTracerTest.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (isaachier) [inline src/jaegertracing/DynamicallyLoadTracerTest.cpp]: 2018

File: src/jaegertracing/DynamicLoadTest.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+TEST(DynamicLoad, validVersion)
+{
+    const void* errorCategory = nullptr;
+    void* tracerFactory = nullptr;
+    const auto rcode = OpenTracingMakeTracerFactory(
+        OPENTRACING_VERSION, &errorCategory, &tracerFactory);
+    ASSERT_EQ(rcode, 0);
+    ASSERT_EQ(errorCategory, nullptr);
+    ASSERT_NE(tracerFactory, nullptr);
+    delete static_cast<opentracing::TracerFactory*>(tracerFactory);
```
Comment (isaachier) [inline src/jaegertracing/DynamicLoadTest.cpp]: Interesting use of `static_cast`. StackOverflow agrees with you here: https://stackoverflow.com/a/941959/1930331.

File: src/jaegertracing/TracerFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (isaachier) [inline src/jaegertracing/TracerFactory.cpp]: 2018

File: src/jaegertracing/TracerFactory.cpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#include "TracerFactory.h"
+
+#include "jaegertracing/Tracer.h"
+
+namespace jaegertracing {
+
+opentracing::expected<std::shared_ptr<opentracing::Tracer>>
+TracerFactory::MakeTracer(const char* configuration,
+                          std::string& errorMessage) const noexcept try {
+    YAML::Node yaml;
```
Comment (isaachier) [inline src/jaegertracing/TracerFactory.cpp]: I think I have the whole YAML dependency wrapped in a macro at the moment. If so, can you add this to wrap the existing code.

```
#ifdef JAEGERTRACING_USE_YAML_CPP
// existing code
#else
// return dummy value or construct config by hand
#endif  // JAEGERTRACING_USE_YAML_CPP
```

File: src/jaegertracing/TracerFactory.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (isaachier) [inline src/jaegertracing/TracerFactory.h]: 2018

File: src/jaegertracing/TracerTest.cpp [hunk 9]
```diff
@@ -122,16 +122,15 @@ TEST(Tracer, testTracer)
...
-    auto tagItr =
-        std::find_if(std::begin(tracer->tags()),
-                     std::end(tracer->tags()),
-                     [](const Tag& tag) {
-                         return tag.key() == kJaegerClientVersionTagKey;
-                     });
+    auto tagItr = std::find_if(
+        std::begin(tracer->tags()),
+        std::end(tracer->tags()),
+        [](const Tag& tag) { return tag.key() == kJaegerClientVersionTagKey; });
```
Comment (isaachier) [inline src/jaegertracing/TracerTest.cpp]: Looks like you ran `clang-format` here. Please run `scripts/clang-format.sh` before pushing next commit to make sure all files are formatted correctly.

File: CMakeLists.txt (hunk 1 above)
Comment (rnburn) [reply CMakeLists.txt]: Done.

File: src/jaegertracing/DynamicLoad.cpp (hunk 2 above)
Comment (rnburn) [reply src/jaegertracing/DynamicLoad.cpp]: Added.

File: src/jaegertracing/DynamicLoad.cpp (hunk 3 above)
Comment (rnburn) [reply src/jaegertracing/DynamicLoad.cpp]: That might make sense.... though it's pretty easy to break the ABI (for example, if you were to change the order of function in the source your vtable entrys could change). It's probably safest (and also easy) to start by requiring a full match.

File: src/jaegertracing/DynamicallyLoadTracerTest.cpp (hunk 4 above)
Comment (rnburn) [reply src/jaegertracing/DynamicallyLoadTracerTest.cpp]: Replaced everywhere.

File: src/jaegertracing/TracerTest.cpp (hunk 9 above)
Comment (rnburn) [reply src/jaegertracing/TracerTest.cpp]: Done.

File: src/jaegertracing/TracerFactory.cpp (hunk 7 above)
Comment (rnburn) [reply src/jaegertracing/TracerFactory.cpp]: I set up dynamic loading to fail if YAML support isn't compiled in and added some test coverage for it as well.

**Comment by rnburn** [comment]
@isaachier - This should be ready for another round of reviewing.

**Comment by isaachier** [comment]
Sorry totally lost track of this. Do you mind changing the copyright from `2018` to `2017-2018` in the files you touched? Otherwise, It looks pretty good to me.

**Comment by isaachier** [comment]
Also, if it isn't too difficult, can you improve coverage of `TracerFactory` by adding an example of a bad config.

**Comment by rnburn** [comment]
@isaachier - There's quite a few tests for invalid configs: https://github.com/rnburn/cpp-client/blob/79b7f6f73091ee8d297b1ae446707ff74e449fcd/src/jaegertracing/TracerFactoryTest.cpp#L22 ... I don't know why they don't seem to be showing up in the coverage.

**Comment by isaachier** [comment]
Hm I'll take a look at build and see whats up

**Comment by rnburn** [comment]
@isaachier replaced 2018 with 2017-2018.

File: src/jaegertracing/DynamicallyLoadTracerTest.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#include <string>
+
+#include <opentracing/dynamic_load.h>
+
+// Verify that Jaeger's shared library can be dynamically loaded and used
+// as a plugin.
+int main(int argc, char* argv[])
+{
+    if (argc != 2) {
+        std::cerr << "DynamicLoadTest <tracer-librar>\n";
```
Comment (isaachier) [inline src/jaegertracing/DynamicallyLoadTracerTest.cpp]: Should be `<tracer-library>`

Comment (rnburn) [reply src/jaegertracing/DynamicallyLoadTracerTest.cpp]: Fixed.

File: src/jaegertracing/LogRecord.cpp [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017 Uber Technologies, Inc.
+ * Copyright (c) 2017-2018 Uber Technologies, Inc.
```
Comment (isaachier) [inline src/jaegertracing/LogRecord.cpp]: I'm sorry to be annoying about this. The idea was to only change the modified files. That way we only update the changed files. See this commit for a related change in the main Jaeger repo: https://github.com/jaegertracing/jaeger/commit/d7d1c74fa74f77d2c7faa50a0c568f979d9a30fd.

Comment (rnburn) [reply src/jaegertracing/LogRecord.cpp]: Is this what you're looking for? The files I added have 2018 and the file that got modified has 2017-2018.

**Pull request merged**